model_training/ml_trainmodel.py

```diff
@@ -226,17 +238,17 @@ def accumulate_asymptotic_loss(dataset_list, traintest):
         print()
         def accumulate_stable_loss(dataset_list, traintest):
             total_loss = torch.tensor(0.0, requires_grad=False)
-            for dataset in dataset_list:
-                F4i = dataset.tensors[0].to(parms["device"])
-                stable_true = dataset.tensors[1].to(parms["device"])
+            with torch.no_grad():
+                for dataset in dataset_list:
+                    F4i = dataset.tensors[0].to(parms["device"])
+                    stable_true = dataset.tensors[1].to(parms["device"])
 
-                _, _, stable_pred = model.predict_all(F4i)
+                    _, _, stable_pred = model.predict_all(F4i)
 
-                print(torch.sum(torch.abs(stable_pred-stable_true)).item()/stable_pred.shape[0],"fractional difference in stable points")
+                    print(torch.sum(torch.abs(stable_pred-stable_true)).item()/stable_pred.shape[0],"fractional difference in stable points")
 
-                # accumulate stability loss using MSE instead of BCE because it is easier to interpret
-                total_loss = total_loss + torch.exp(-model.log_task_weights["stability"] ) * \
-                    contribute_loss(stable_pred, stable_true, traintest, "stability", comparison_loss_fn)
+                    total_loss = total_loss + torch.exp(-model.log_task_weights["stability"] ) * \
+                        contribute_loss(stable_pred, stable_true, traintest, "stability", comparison_loss_fn)
```

Comment:
Similarly, I think it would be better to have no_grad where the function is used.